paramiko/auth_handler.py

```diff
@@ -243,13 +343,51 @@ def _parse_service_accept(self, m):
                 password = bytestring(self.password)
                 m.add_string(password)
             elif self.auth_method == 'publickey':
-                m.add_boolean(True)
-                m.add_string(self.private_key.get_name())
-                m.add_string(self.private_key)
-                blob = self._get_session_blob(
-                    self.private_key, 'ssh-connection', self.username)
-                sig = self.private_key.sign_ssh_data(blob)
-                m.add_string(sig)
+                if self.pkcs11pin is None:
+                    # Private Key
+                    m.add_boolean(True)
+                    m.add_string(self.private_key.get_name())
+                    m.add_string(self.private_key)
+                    blob = self._get_session_blob(
+                        self.private_key, 'ssh-connection', self.username)
+                    sig = self.private_key.sign_ssh_data(blob)
+                    m.add_string(sig)
+                else:
+                    # Smartcard PKCS11 Private Key
+                    fields = self._pkcs11_get_public_key().split(' ')
+
+                    if len(fields) < 2:
+                        SSHException("Not enough fields found in pkcs11 key")
+
+                    keytype = fields[0]
+                    pub_key = fields[1]
+
+                    try:
+                        pub_key = b(pub_key)
+                        if keytype == 'ssh-rsa':
+                            pub_key = RSAKey(data=decodebytes(pub_key))
+                        elif keytype == 'ssh-dss':
+                            pub_key = DSSKey(data=decodebytes(pub_key))
+                        elif keytype in ECDSAKey\
+                            .supported_key_format_identifiers():
+                            pub_key = ECDSAKey(data=decodebytes(pub_key),
+                                               validate_point=False)
+                        elif keytype == 'ssh-ed25519':
+                            pub_key = Ed25519Key(data=decodebytes(pub_key))
```

Comment:
I think this "if keytype == blah: key = blah(data)" pattern can be put in a helper function somewhere such that hostkeys.py could share use of it.